### PR TITLE
[skrifa] autohint: point analysis

### PR DESCRIPTION
**Pull request by dfrg**
Implements the initial point analysis common to all writing systems:
1. Detect subsequent points within a heuristic threshold and mark them near
2. Compute dominant input/output directions for each point pair, skipping intermediate sequences of near points
3. Simplify, marking directionless points as weak
4. Identify remaining weak interpolation points

### CODE REVIEW

File: skrifa/src/outline/autohint/cycling.rs [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+            self.last
+        } else {
+            ix - 1
+        }
+    }
+}
+
+/// Iterator that begins at `start + 1` and cycles through all items
+/// of the slice in forward order, ending with `start`.
+pub(super) fn cycle_forward<T>(items: &[T], start: usize) -> impl Iterator<Item = (usize, &T)> {
```
[the remark this answers is not in the document]
Comment (rsheeter) [reply skrifa/src/outline/autohint/cycling.rs:33]: +1 to this concurn. Aside: I wish "non-empty slice/vec" was readily expressed in the type system.

File: skrifa/src/outline/autohint/cycling.rs [hunk 2]
```diff
@@ -0,0 +1,79 @@
+//! Various helpers for dealing with iteration of a slice
+//! that represents a loop. Specifically, outline contours.
+
+#[derive(Copy, Clone)]
+pub(super) struct IndexCycler {
+    last: usize,
+}
+
+impl IndexCycler {
+    pub fn new(len: usize) -> Self {
```
[the remark this answers is not in the document]
Comment (dfrg) [reply skrifa/src/outline/autohint/cycling.rs]: Feels bad indeed. Changed to return an option, yielding `None` when len is 0.

File: skrifa/src/outline/autohint/cycling.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (dfrg) [reply skrifa/src/outline/autohint/cycling.rs:38]: The iterator is based on (0..len) so the modulo is never executed when the slice length is 0. Added code to tests to prove these don't panic on empty slices.

File: skrifa/src/outline/autohint/cycling.rs [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    let start = start + 1;
+    (0..len).map(move |ix| {
+        let real_ix = (ix + start) % len;
+        (real_ix, &items[real_ix])
+    })
+}
+
+/// Iterator that begins at `start - 1` and cycles through all items
+/// of the slice in reverse order, ending with `start`.
+pub(super) fn cycle_backward<T>(items: &[T], start: usize) -> impl Iterator<Item = (usize, &T)> {
```
[the remark this answers is not in the document]
Comment (dfrg) [reply skrifa/src/outline/autohint/cycling.rs:49]: I prefer keeping the free functions for now but I'll consider this in the future when I have a better understanding of all the iteration patterns in the code

File: skrifa/src/outline/autohint/outline.rs [hunk 4]
```diff
@@ -135,6 +146,213 @@ impl Outline {
...
+            }
+            let mut prev_ix = points.len() - 1;
+            let mut ix = 0;
+            while ix < points.len() {
+                let point = points[ix];
+                let prev = &mut points[prev_ix];
+                // <https://gitlab.freedesktop.org/freetype/freetype/-/blob/57617782464411201ce7bbc93b086c1b4d7d84a5/src/autofit/afhints.c#L1017>
+                let out_x = point.fx - prev.fx;
+                let out_y = point.fy - prev.fy;
+                if out_x.abs() + out_y.abs() < near_limit {
```
[the remark this answers is not in the document]
Comment (dfrg) [reply skrifa/src/outline/autohint/outline.rs:166]: The function would be something like `font_unit_taxi_cab_distance` 🤷 due to the computation and the fact that the point type will grow to include 2 more coordinate sets (scaled and hinted). Also, the third use of this computation accumulates the distances for each axis independently so the function wouldn't be useful there.

File: skrifa/src/outline/autohint/outline.rs [hunk 5]
```diff
@@ -135,6 +146,213 @@ impl Outline {
...
+                    if point.out_dir != Direction::None {
+                        // Point lies on a vertical or horizontal segment but
+                        // not at start or end
+                        make_weak = true;
+                    } else {
+                        let u_index = point.u as usize;
+                        let v_index = point.v as usize;
+                        let next_u = points[u_index];
+                        let prev_v = points[v_index];
+                        if is_corner_flat(
```
[the remark this answers is not in the document]
Comment (dfrg) [reply skrifa/src/outline/autohint/outline.rs:306]: I'll circle back to this if it turns out this is the only use of `is_corner_flat` or that all uses operate on unmodified font unit coords of `Point`s.

**Pull request merged**